### ISSUE TITLE
Resolve shortlinks on the server side to improve privacy

### ISSUE THREAD

**Issue by allo-** [mastodon/mastodon#23361]
### Pitch

Shortening links is not necessary on mastodon because every URL counts for the same number of characters. But there are other use cases for shortlinks that are useful to the author but detrimental to the reader, such as hiding referral links and adding analytics.

Examples:
1) https://short1.example.com/abc -> https://thisisalonglink.example.com/with/a/long/path
2) https://short2.example.com/123 -> https://affliate.bonus.example.com/123 (setting a referral cookie) -> https://example.com
3) https://short3.example.com/xyz (storing referer, user-agent and other client characteristics in a database) -> https://normal-link.example.com

In the first case, the shortlink is unnecessary, but does no harm.  
In the second case, a person can distribute spam links that only become obvious after clicking (e.g. recognisable by a ?refid=xyz parameter in the final url) or can even be hidden by storing the refid and immediately redirecting to the target site without any visible referral ids.
The latter is only a privacy issue. It does not help spammers directly, but it does help them collect data about users who click on the link.

Proposed solution:
Mastodon already fetches links to create previews. When fetching the content, it can replace the original link with the final target of the redirect chain. This will either reveal that there is a ?refid=xy parameter, or if there is another redirect, it will only use the final URL. Any redirects that are only used for analytics will also be removed, giving users more privacy.

There are two places where this could happen:
1) If there is no malicious intent, it can be handled on the author's instance. For example, if someone is using a service that schedules content, and the scheduling service uses shortlinks, the author's instance can already strip the shortlink.
2) Alternatively, assuming that malicious authors would choose instances that do not clean their links, the receiving instances could replace the link. Since the instances are already fetching the content to create previews, and do not trust the author's instance to include a valid preview, they can also replace the link with the final redirect target.

### Motivation

As Mastodon grows in popularity, there are now accounts trying to use the same tactics for analytics and referral bonuses as they do on Twitter. This is done both manually (people creating links on the site of their favourite shortlink service with analytics) and automatically, i.e. through tools for automatic posting and/or scheduling of toots, which also provide users with click analytics by using their own shortlink service.

While these analytics are a benefit to the author, they are often to the detriment of the user who clicks on the link. They are also part of a common scheme used by spammy accounts on Twitter to mix funny content with commercial links, which are often deleted after a few minutes when enough of the audience has clicked on the link.

**Comment by untitaker** [comment]
redirects are already resolved for link previews, and sometimes that comes with somewhat inconvenient side-effects. to test this out, you can try these steps:

1. register an account on a mastodon instance hosted in germany
2. post a link to a techcrunch article
3. the resulting post will have a link to techcrunch.com, but the link preview will link directly to consent.yahoo.com

![image](https://user-images.githubusercontent.com/837573/216156098-fa4a0c7d-b5dc-4c2b-a9de-783c7838d77f.png)

that link to consent.yahoo.com is not usable at all. if mastodon resolved redirects for all links as you propose, a post like the above would have no functioning links to techcrunch at all.

i understand that the link preview has to resolve redirects in order to show useful title images, but there's too much breakage that can occur if your proposal is done indiscriminately for all kinds of URLs. so if this lands I believe it should not be enabled by default, or only done for a handful of domains. and if it is done only for a few domains I am not sure it is worth the effort at all.

**Comment by allo-** [comment]
This feature could be combined with a whitelist/blacklist approach if necessary. There could also be a "show original url" button somewhere. The point is that there are some easily trackable redirect services as well as services designed specifically to track individual users. Also, the shadier reflink sites often provide links that do not look like they contain a refid, so the user can post it without anyone suspecting that it is a paid link.

There are already some client-side options like [UntrackMe](https://f-droid.org/packages/app.fedilab.nitterizeme/), but having the server pre-process it would have the advantage that links that really need to be loaded to extract the final URL would be fetched by the server and not by the user, who even without permanent cookies would at least give away the combination of refid and IP address.

**Comment by mcclure** [comment]
Sometimes, I will use a link shortener because the potential to move a link after the fact is useful. For example, maybe I have content hosted somewhere provisionally but I expect I might move it later. I am not generally gathering analytics (I don't think I know how to do that) so I am not engaging in the bad behavior you are trying to stop, but if you added this feature this would potentially break any "shortener as jump-pad to current location" links I posted, and I might not realize you had done this or know how to fix it (because I don't know what you did).

Now you might say that posting a link that later redirects to somewhere totally different is also bad behavior but the point is silently editing a user's posts to change the link they are linking to can have unintended side effects.

**Comment by untitaker** [comment]
I think this is a great point. A concrete example is using a custom domain together with linktr.ee. The only way to make use of custom domains in that context is to redirect (which many registrars offer OOTB)

**Comment by spirillen** [comment]
TL;DR: I agree — almost any short URL is used for tracking. Regardless of the backend, when the destination is resolved it triggers tracking, and as long as these short, opaque URLs exist users click them and are tracked without consent. These links can propagate through other software too.

For example, `https://mastodon.world/@ionanthosphotography/110939562508751366` links to a `bit.ly` URL (see https://0xacab.org/my-privacy-dns/matrix/-/issues/178475); without proper protections, users’ privacy is compromised.

I fully support this suggestion. Of course, three lists should be maintained: whitelist, greylIst and blacklist.

**Comment by shaedrich** [comment]
> 3. the resulting post will have a link to techcrunch.com, but the link preview will link directly to consent.yahoo.com
> 
> 
> ![image](https://user-images.githubusercontent.com/837573/216156098-fa4a0c7d-b5dc-4c2b-a9de-783c7838d77f.png)
> 
> that link to consent.yahoo.com is not usable at all. if mastodon resolved redirects for all links as you propose, a post like the above would have no functioning links to techcrunch at all.

Well, that is not the problem we discuss in this issue. If you think, this is a bug, you should report this for the link preview cards
